grammar.js

```diff
@@ -588,27 +588,26 @@ module.exports = grammar({
     _decimal: ($) => /[0-9][0-9_]*/,
     _octal: ($) => /0[oO][0-7_]+/,
     _binary: ($) => /0[bB][0-1_]+/,
-    _bit_string_segment_option_unit: ($) =>
-      seq("unit", "(", alias($.integer, $.bit_string_segment_option_unit), ")"),
-    _bit_string_segment_option_literal: ($) =>
+    _bit_string_segment_option: ($) =>
       choice(
-        alias("binary", $.bit_string_segment_option_binary),
-        alias("bytes", $.bit_string_segment_option_binary),
-        alias("int", $.bit_string_segment_option_int),
-        alias("float", $.bit_string_segment_option_float),
-        alias("bit_string", $.bit_string_segment_option_bit_string),
-        alias("bits", $.bit_string_segment_option_bit_string),
-        alias("utf8", $.bit_string_segment_option_utf8),
-        alias("utf16", $.bit_string_segment_option_utf16),
-        alias("utf32", $.bit_string_segment_option_utf32),
-        alias("utf8_codepoint", $.bit_string_segment_option_utf8_codepoint),
-        alias("utf16_codepoint", $.bit_string_segment_option_utf16_codepoint),
-        alias("utf32_codepoint", $.bit_string_segment_option_utf32_codepoint),
-        alias("signed", $.bit_string_segment_option_signed),
-        alias("unsigned", $.bit_string_segment_option_unsigned),
-        alias("big", $.bit_string_segment_option_big),
-        alias("little", $.bit_string_segment_option_little),
-        alias("native", $.bit_string_segment_option_native)
+        "binary",
+        "bytes",
+        "int",
+        "float",
+        "bit_string",
+        "bits",
+        "utf8",
+        "utf16",
+        "utf32",
+        "utf8_codepoint",
+        "utf16_codepoint",
+        "utf32_codepoint",
+        "signed",
+        "unsigned",
+        "big",
+        "little",
+        "native",
+        seq("unit", "(", $.integer, ")")
```

Comment:
I'm not sure this is a good idea but I was just thinking it might be nice to have this be `($) => /[\w\d_]+/` or something similarly permissive and then to capture these as function.capture/keyword/etc in the queries

```
((bit_string_segment_option) @function.builtin)
 (#match? @function.builtin "^(binary|bytes|int|float|...)$"))
```

with the idea being that the parser always returns valid nodes but they're only highlighted when they're known bit-string options.

I've been considering doing the same for [tree-sitter-git-rebase](https://github.com/the-mikedavis/tree-sitter-git-rebase) because it feels "jumpy" when you write out `pick` for example (`p` is highlighted red, then it's unhighlighted with `pi`, so on until `pick` is highlighted again).

OTOH it feels a bit weird to extract this from the parser and put it in queries :thinking:

Reply:
To your point, I think this does raise a fundamental question of "what checks belong in the parser vs a later step (e.g. semantic analyzer)?". The approach that I took was based on the Gleam compiler's parser, which rather specific parsing mechanisms, which could be construed as strict checks.

That said, I can very much see the benefit of having the parser do less checking/accept some invalid code as with the example you outlined. I think for now I'll opt to keep this as-is, but going forward we may reconsider :grin: